src/server/game/Handlers/MiscHandler.cpp

```diff
@@ -1448,16 +1448,27 @@ void WorldSession::HandleTimeSyncResp(WorldPacket& recvData)
     recvData >> counter >> clientTicks;
 
     if (counter != _player->m_timeSyncCounter - 1)
+    {
         TC_LOG_DEBUG("network", "Wrong time sync counter from player %s (cheater?)", _player->GetName().c_str());
+        // todo: send a new one?
+        return;
+    }
 
-    TC_LOG_DEBUG("network", "Time sync received: counter %u, client ticks %u, time since last sync %u", counter, clientTicks, clientTicks - _player->m_timeSyncClient);
-
-    uint32 ourTicks = clientTicks + (getMSTime() - _player->m_timeSyncServer);
-
-    // diff should be small
-    TC_LOG_DEBUG("network", "Our ticks: %u, diff %u, latency %u", ourTicks, ourTicks - clientTicks, GetLatency());
+    // time it took for the request to travel to the client, for the client to process it and reply and for response to travel back to the server.
+    uint32 roundTripDuration = getMSTimeDiff(_player->m_timeSyncServer, getMSTime());
+    uint32 lagDelay = roundTripDuration / 2; // we assume that the request processing time equals 0.
 
-    _player->m_timeSyncClient = clientTicks;
+    /*
+    clockDelta = serverTime - clientTime
+    where
+    serverTime: time that was displayed on the cloak of the SERVER at the moment when the client processed the SMSG_TIME_SYNC_REQUEST packet.
```

Comment:
cloak = a cape or drape, a sleeveless outdoor overgarment that hangs loosely from the shoulders.
clock = an instrument to indicate, keep, and co-ordinate time.